### DISCUSSION TITLE
Sprite max size?

### DISCUSSION THREAD

**Discussion by PabloAlvarezCalvo** [Bodmer/TFT_eSPI#2280] [Q&A - Sprites]
Been working with this library without issues for ST7789 displays in 240x240 pixel format without issue. Today I got a new 240x320 2" display (which also uses a ST7789 driver). I'm working with a Wemos Lolin 32.

Working with sprites, previous sketches work without issue. In previous sketches defined a 240x240 sprite which I push continuously. With this new display I can push the sprite with (0, 80) offset and verify I can use the whole screen. However, if I define a 240 by 320 sprite the sketch stops working. Is there any size/memory limit?

Code is this simple:

    #include <SPI.h>
    
    #include <TFT_eSPI.h>
    TFT_eSPI tft = TFT_eSPI();   // Invoke library
    TFT_eSprite img = TFT_eSprite(&tft);
    
    //#define USERCOLOUR TFT_CYAN
    //<-----------------SETUP------------------->
    void setup() {
        
        tft.begin();     // initialize a ST7789 chip
        tft.setSwapBytes(true); // Swap the byte order for pushImage() - corrects endianness
        
        tft.fillScreen(TFT_BLACK);
        tft.setRotation(0);

        img.createSprite(240, 250); //This works       
        //img.createSprite(240, 320); Blank screen
        img.fillSprite(TFT_BLACK); 
    
    }
    
    //<------------------LOOP------------------->
    void loop(){
    
        img.drawRect(0, 0, 240, 240, TFT_WHITE);
        img.pushSprite(0, 0);
    }

And this is my setup

    #define ST7789_DRIVER     // Configure all registers
    
    #define TFT_WIDTH  240
    #define TFT_HEIGHT 320
    
    #define CGRAM_OFFSET      // Library will add offsets required
    
    //#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
    //#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red
    
    //#define TFT_INVERSION_ON
    //#define TFT_INVERSION_OFF
    
    // DSTIKE stepup
    //#define TFT_DC    23
    //#define TFT_RST   32
    //#define TFT_MOSI  26
    //#define TFT_SCLK  27
    
    // Generic ESP32 setup
    #define TFT_MISO -1 // Not connected
    #define TFT_MOSI 23
    #define TFT_SCLK 18
    #define TFT_CS   16 //Sólo pre-Beskar
    #define TFT_DC    2
    #define TFT_RST   4  // Connect reset to ensure display initialises
    
    // For NodeMCU - use pin numbers in the form PIN_Dx where Dx is the NodeMCU pin designation
    //#define TFT_CS   -1      // Define as not used
    //#define TFT_DC   PIN_D1  // Data Command control pin
    //#define TFT_RST  PIN_D4  // TFT reset pin (could connect to NodeMCU RST, see next line)
    //#define TFT_RST  -1    // TFT reset pin connect to NodeMCU RST, must also then add 10K pull down to TFT SCK
    
    
    #define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
    #define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
    #define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
    #define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
    #define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
    #define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
    //#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
    #define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
    
    #define SMOOTH_FONT
    
    
    // #define SPI_FREQUENCY  27000000
    #define SPI_FREQUENCY  40000000
    
    #define SPI_READ_FREQUENCY  20000000
    
    #define SPI_TOUCH_FREQUENCY  2500000
    
    // #define SUPPORT_TRANSACTIONS

**Comment by PabloAlvarezCalvo** [comment]
After re-reading the Readme (what a concept lol) I see that my problem is the limit of the RAM, so I'll find a workaround.

**Reply by annguyen18** [reply]
Hi, sorry if I bother you, I have a question, the wemos ESP32 has 520 Kbytes of RAM, (I don't know how many bits colors did you use so I assume you used 2 bits), 320*240*2=153600 bytes ~ 154kbytes, so how is 520 Kbytes not enough for 154 kbytes? 
I figured you have found a workaround?
really appreciate if you reply to this question.

**Reply by marcel-licence** [reply]
I was facing the same issue. Seems that something consumes to much memory. I tried to implement a static memory allocation but using
`static uint8_t spriteBuffer[320*240*2 + 2];`
caused an overflow of ~50k;

Compiling the following code:
```
#include <TFT_eSPI.h>                 // Include the graphics library (this includes the sprite functions)
TFT_eSPI tft = TFT_eSPI(); // Create object "tft"
TFT_eSprite scr = TFT_eSprite(&tft); // Create Sprite object "img" with pointer to "tft" object

uint8_t staticSprite[240 * 320 * 2 + 2];

void setup(void)
{
    tft.init();
    tft.setRotation(0);
    Serial.begin(57600); 
    tft.fillScreen(0);
    scr.createSpriteStatic(240, 320, 1, staticSprite);
}

void loop(void)
{

}
```
results in:
```
[...]/xtensa-esp32-elf/bin/ld.exe: [...]/test.ino.elf section `.dram0.bss' will not fit in region `dram0_0_seg'
[...]/xtensa-esp32-elf/bin/ld.exe: DRAM segment data does not fit.
[...]/xtensa-esp32-elf/bin/ld.exe: DRAM segment data does not fit.
[...]/xtensa-esp32-elf/bin/ld.exe: region `dram0_0_seg' overflowed by 51248 bytes
```
For testing I've created a function createSpriteStatic which uses the static buffer instead of calling calloc during runtime.

The selected Display is:
`#include <User_Setups/Setup25_TTGO_T_Display.h>    // Setup file for ESP32 and TTGO T-Display ST7789V SPI bus TFT`

**Reply by IanBUK** [reply]
The code I’m working on uses a Sparkfun Thing esp32 and a 800x480 ssd1963. I couldn’t make sprites bigger than about 200x200. So, I create multiple sprites and tile them. I’ve had to drop the colour depth to 8 bit, but it’s given me a 400x400 region of the screen covered by sprites. The refresh rate is just under double that of straight drawing, with no optimisations for tiling. If you look closely, you occasionally see a bit of tearing, but it’s fine.
